### DISCUSSION TITLE
Timeouts in Google Cloud Run

### DISCUSSION THREAD

**Discussion by jarshwah** [fastapi/fastapi#9000] [Questions]
I implemented a redirect service using Fast API on Google Cloud Run. Unfortunately, the health checks succeeded only rarely and the service would die. After changing over to a flask service the timeout issues went away.

I'm raising this issue in the hopes that either:

1. Someone can point out where I was being not very smart; or
2. Someone else seeing similar behaviour has a place to discuss their issues

I suspect it'll be 1 :)

### Example

Here's a trimmed down repo, including all of the production configuration, that was exhibiting the issue. I've only removed the core business logic.

https://github.com/jarshwah/fastapi-timeout-gcp

```python
import logging.config

from fastapi import FastAPI, Request, responses, status
from fastapi.exceptions import HTTPException
from utils import configure_logging, is_valid

configure_logging()
log = logging.getLogger("direktor")
app = FastAPI()


@app.get("/health")
async def health_check():
    return {"healthy": True}


@app.get("/{path:path}")
async def oauth_redirect(request: Request, path: str, state: str = ""):
    target_domain, _, _ = state.partition("|")
    path = f"/{path}"
    if not is_valid(target_domain, path):
        log.info("unauthorized domain=%s path=%s", target_domain, path)
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST)

    full_url = f"https://{target_domain}{path}"
    return responses.RedirectResponse(full_url, status_code=302)


@app.exception_handler(Exception)
async def exception(request: Request, exc: Exception):
    log.exception(exc)
    return responses.JSONResponse(
        status_code=500, content={"message": "An error occurred. Please try again later."}
    )


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=9050)
```

In this example the health check would be hit every 1 minute (or so) from an uptime monitor service. This should have been enough to keep the container "warm" for fast initial requests.

The Dockerfile used starts uvicorn with 10 workers:

```Dockerfile
FROM python:3.8-slim
RUN pip install poetry
WORKDIR /app
COPY poetry.lock pyproject.toml /app/
ENV POETRY_VIRTUALENVS_CREATE false
RUN poetry install --no-dev
COPY main.py /app/main.py
COPY utils.py /app/utils.py
CMD uvicorn main:app --workers 10 --port $PORT --host 0.0.0.0
```

And the cloud build definition specifies a concurrency of 10 (10 concurrent requests per container) and a max instances of 20 (20 total containers) for an estimated total capacity of 200 concurrent requests.

### Problem

The issue manifested as the uptime service reporting constant failures, and downtime measured in hours. Sometimes the service would seem to run fine for a few hours and other times it'd die immediately.

In hindsight, I think my issue was likely starting uvicorn with 10 workers thinking of it as a queue capacity rather than 10 individual processes all competing for very few cpu resources. We did load test this service on our local machines and found that more workers performed better, but our local machines are also considerably higher spec than the provisioned container servers.

### Possible Solutions

The [Deployment Docs](https://fastapi.tiangolo.com/deployment/) are excellent, but they could probably benefit from some "deploying into production best practises" that involve configuring the application servers in a responsible manner. These are always the most difficult docs to find - as they are plentiful, opinionated, and often wrong.

**Comment by insomnes** [comment]
@jarshwah hello!
I'm not an expert, but maybe you could try run it with gunicorn + uvicorn instead of uvicorn workers
https://www.uvicorn.org/#running-with-gunicorn

You can also use FastAPI docker image https://github.com/tiangolo/uvicorn-gunicorn-fastapi-docker
as described at 
https://fastapi.tiangolo.com/deployment/docker/

**Comment by jybaek** [comment]
Have you checked the stackdriver log?
